src/defck.rs

```diff
@@ -91,6 +92,77 @@ fn check_hyp(
     check(db, addr, f)
 }
 
+fn check_lhs(syntax: Label, f: SubFormulaRef<'_>, params: &mut HashSet<Label>) -> bool {
+    syntax == f.label()
+        && f.children()
+            .all(|v| v.is_variable() && params.insert(v.label()))
+}
+
+fn find_lhs<'a>(
+    syntax: Label,
+    f: SubFormulaRef<'a>,
+    lhs: &mut Option<SubFormulaRef<'a>>,
+    params: &mut HashSet<Label>,
+) -> bool {
+    if f.label() == syntax {
+        if let Some(lhs2) = lhs {
+            f == *lhs2
+        } else {
+            check_lhs(syntax, f, params) && {
+                *lhs = Some(f);
+                true
+            }
+        }
+    } else {
+        f.children().all(|f| find_lhs(syntax, f, lhs, params))
+    }
+}
+
+struct JustificationSubst<'a> {
+    vars: HashMap<Label, SubFormulaRef<'a>>,
+    rhs: Option<SubFormulaRef<'a>>,
+}
+
+fn _check_rhs(
+    syntax: Label,
+    f: SubFormulaRef<'_>,
+    params: &HashSet<Label>,
+    vars: &HashMap<Label, Label>,
+) -> bool {
+    f.labels_iter().all(|(label, var)| {
+        label != syntax && (!var || params.contains(&label) || !vars.contains_key(&label))
+    })
+}
+fn match_justification<'a>(
+    syntax: Label,
+    just: SubFormulaRef<'a>,
+    f: SubFormulaRef<'a>,
+    subst: &mut JustificationSubst<'a>,
+) -> bool {
+    if f.label() == syntax {
+        if let Some(rhs2) = &subst.rhs {
+            just == *rhs2
+        } else {
+            subst.rhs = Some(just);
+            true
+        }
+    } else if just.is_variable() {
+        match subst.vars.entry(just.label()) {
+            Entry::Occupied(e) => *e.get() == f,
+            Entry::Vacant(e) => {
+                e.insert(f);
+                true
+            }
+        }
+    } else {
+        just.label() == f.label()
+            && just
+                .children()
+                .zip(f.children())
+                .all(|(just, f)| match_justification(syntax, just, f, subst))
+    }
+}
```

Comment:
Couldn't we simply use `Formula::unify` ?

Reply:
Almost; it's a unification which handles the definition subterm specially